src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -1764,6 +1790,49 @@ ReadOnlyCollection<DbColumn> IDbColumnSchemaGenerator.GetColumnSchema()
             if (FieldCount == 0) // No resultset
                 return null;
 
+            var table = GetEmptySchemaTable();
+
+            foreach (var column in GetColumnSchema())
+            {
+                AddColumnToSchemaTable(table, column);
+            }
+
+            return table;
+        }
+
+        /// <summary>
+        /// Asynchronously returns a System.Data.DataTable that describes the column metadata of the DataReader.
+        /// </summary>
+#if !NET461 && !NETSTANDARD2_0 && !NETSTANDARD2_1 && !NETCOREAPP3_0
+        public override Task<DataTable?> GetSchemaTableAsync(CancellationToken cancellationToken = default)
+#else
+        public Task<DataTable?> GetSchemaTableAsync(CancellationToken cancellationToken = default)
+#endif
+        {
+            if (cancellationToken.IsCancellationRequested)
+                return Task.FromCanceled<DataTable?>(cancellationToken);
+
+            using (NoSynchronizationContextScope.Enter())
+                return GetSchemaTable(cancellationToken);
+        }
+
+        async Task<DataTable?> GetSchemaTable(CancellationToken cancellationToken = default)
```

Comment:
Should be named GetSchemaTableAsync, no?

Also, this seems to be called by GetSchemaTableAsync - which checks the cancellation token and applies NoSynchronizationContextScope - but then calls GetColumnSchemaAsync, which does that again. It also duplicates the sync version.

The easiest here would be for the top-level, public GetSchemaTable methods to both call into a unified, internal GetSchemaTable method which accepts an async flag, and then calls the internal GetColumnSchema which also accepts an async flag.